File: packages/provide-uterm-go/vt/edit.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package vt

// Drawing, scrolling, and line/character insertion, deletion and erasure.
// Each method mirrors the corresponding pyte.screens.Screen handler,
// including its sparse-buffer quirks: erase operations rewrite only cells
// that are explicitly stored, and scroll/shift operations move whole line
// maps around.

// Draw displays characters at the cursor position, advancing the cursor
// and wrapping at the right margin when DECAWM is set. Zero-width
// combining characters merge into the preceding cell; other unprintable
// characters abort the rest of the batch, exactly like pyte.
func (s *Screen) Draw(data string) {
	cs := s.g0Charset
	if s.charset == 1 {
		cs = s.g1Charset
	}

	for _, char := range data {
		if char >= 0 && char < 256 {
			char = cs[char]
		}
		charWidth := runeWidth(char)

		// If this was the last column in a line and auto wrap mode is
		// enabled, move the cursor to the beginning of the next line,
		// otherwise replace characters already displayed.
		if s.cursor.X == s.columns {
			if s.mode[DECAWM] {
				s.CarriageReturn()
				s.LineFeed()
			} else if charWidth > 0 {
				s.cursor.X -= charWidth
			}
		}

		// In insert mode new characters shift old ones to the right.
		if s.mode[IRM] && charWidth > 0 {
			s.InsertCharacters(charWidth)
		}

		line := s.line(s.cursor.Y)
		switch {
		case charWidth == 1:
			line[s.cursor.X] = withData(s.cursor.Attrs, string(char))
		case charWidth == 2:
			// A two-cell character has a stub slot after it.
			line[s.cursor.X] = withData(s.cursor.Attrs, string(char))
			if s.cursor.X+1 < s.columns {
				line[s.cursor.X+1] = withData(s.cursor.Attrs, "")
			}
		case charWidth == 0 && combiningClass(char) > 0:
			// A zero-cell character combines with the previous character
			// either on this or the preceding line.
			if s.cursor.X > 0 {
				last := cellAt(line, s.cursor.X-1, s.defaultChar())
				normalized := nfcNormalize(last.Data + string(char))
				line[s.cursor.X-1] = withData(last, normalized)
			} else if s.cursor.Y > 0 {
				prev := s.line(s.cursor.Y - 1)
				last := cellAt(prev, s.columns-1, s.defaultChar())
				normalized := nfcNormalize(last.Data + string(char))
				prev[s.columns-1] = withData(last, normalized)
			}
		default:
			// Unprintable character, or one that does not advance the
			// cursor: stop processing the rest of the batch.
			return
		}

		if charWidth > 0 {
			s.cursor.X = min(s.cursor.X+charWidth, s.columns)
		}
	}
}

// Index moves the cursor down one line in the same column; at the bottom
// margin the scrolling region scrolls up.
func (s *Screen) Index() {
	top, bottom := s.marginsOrScreen()
	if s.cursor.Y == bottom {
		for y := top; y < bottom; y++ {
			if src, ok := s.buffer[y+1]; ok {
				s.buffer[y] = src
			} else {
				s.buffer[y] = make(map[int]Char)
			}
		}
		delete(s.buffer, bottom)
	} else {
		s.CursorDown(0)
	}
}

// ReverseIndex moves the cursor up one line in the same column; at the
// top margin the scrolling region scrolls down.
func (s *Screen) ReverseIndex() {
	top, bottom := s.marginsOrScreen()
	if s.cursor.Y == top {
		for y := bottom; y > top; y-- {
			if src, ok := s.buffer[y-1]; ok {
				s.buffer[y] = src
			} else {
				s.buffer[y] = make(map[int]Char)
			}
		}
		delete(s.buffer, top)
	} else {
		s.CursorUp(0)
	}
}

// LineFeed performs an Index and, when LNM is set, a carriage return.
func (s *Screen) LineFeed() {
	s.Index()
	if s.mode[LNM] {
		s.CarriageReturn()
	}
}

// InsertLines inserts count blank lines at the cursor line; lines at and
// below the cursor move down, and lines pushed past the bottom margin are
// lost. A noop when the cursor is outside the scrolling region.
func (s *Screen) InsertLines(count int) {
	if count == 0 {
		count = 1
	}
	top, bottom := s.marginsOrScreen()

	if top <= s.cursor.Y && s.cursor.Y <= bottom {
		for y := bottom; y >= s.cursor.Y; y-- {
			if y+count <= bottom {
				if src, ok := s.buffer[y]; ok {
					s.buffer[y+count] = src
				}
			}
			delete(s.buffer, y)
		}
		s.CarriageReturn()
	}
}

// DeleteLines deletes count lines starting at the cursor line; lines
// below move up. A noop when the cursor is outside the scrolling region.
//
// Quirk preserved from pyte: a destination line is only replaced when the
// source line is explicitly stored in the sparse buffer, so deleting into
// never-touched lines leaves the destination content in place.
func (s *Screen) DeleteLines(count int) {
	if count == 0 {
		count = 1
	}
	top, bottom := s.marginsOrScreen()

	if top <= s.cursor.Y && s.cursor.Y <= bottom {
		for y := s.cursor.Y; y <= bottom; y++ {
			if y+count <= bottom {
				if src, ok := s.buffer[y+count]; ok {
					delete(s.buffer, y+count)
					s.buffer[y] = src
				}
			} else {
				delete(s.buffer, y)
			}
		}
		s.CarriageReturn()
	}
}

// InsertCharacters inserts count blank characters at the cursor position,
// shifting the rest of the line right; the cursor does not move.
func (s *Screen) InsertCharacters(count int) {
	if count == 0 {
		count = 1
	}
	line := s.line(s.cursor.Y)
	def := s.defaultChar()
	for x := s.columns; x >= s.cursor.X; x-- {
		if x+count <= s.columns {
			line[x+count] = cellAt(line, x, def)
		}
		delete(line, x)
	}
}

// DeleteCharacters deletes count characters starting at the cursor
// position; characters to the right move left with their attributes.
func (s *Screen) DeleteCharacters(count int) {
	if count == 0 {
		count = 1
	}
	line := s.line(s.cursor.Y)
	for x := s.cursor.X; x < s.columns; x++ {
		if x+count <= s.columns {
			if src, ok := line[x+count]; ok {
				delete(line, x+count)
				line[x] = src
			} else {
				line[x] = s.defaultChar()
			}
		} else {
			delete(line, x)
		}
	}
}

// EraseCharacters erases count characters starting at the cursor
// position, setting them to the cursor attributes; the cursor stays put.
func (s *Screen) EraseCharacters(count int) {
	if count == 0 {
		count = 1
	}
	line := s.line(s.cursor.Y)
	for x := s.cursor.X; x < min(s.cursor.X+count, s.columns); x++ {
		line[x] = s.cursor.Attrs
	}
}

// EraseInLine erases the current line: 0 from the cursor to the end, 1
// from the beginning to the cursor, 2 the complete line. Erased cells take
// the cursor attributes. The private flag is accepted for stream
// compatibility and ignored, as in pyte.
func (s *Screen) EraseInLine(how int, private bool) {
	_ = private
	var start, end int
	switch how {
	case 0:
		start, end = s.cursor.X, s.columns
	case 1:
		start, end = 0, s.cursor.X+1
	case 2:
		start, end = 0, s.columns
	default:
		return
	}

	line := s.line(s.cursor.Y)
	for x := start; x < end; x++ {
		line[x] = s.cursor.Attrs
	}
}

// EraseInDisplay erases the screen: 0 from the cursor to the end, 1 from
// the beginning to the cursor, 2 or 3 the whole display (the cursor does
// not move).
//
// Quirk preserved from pyte: on full lines only explicitly stored cells
// are rewritten with the cursor attributes; unset cells keep rendering as
// the default character.
func (s *Screen) EraseInDisplay(how int) {
	var start, end int
	switch how {
	case 0:
		start, end = s.cursor.Y+1, s.lines
	case 1:
		start, end = 0, s.cursor.Y
	case 2, 3:
		start, end = 0, s.lines
	default:
		return
	}

	for y := start; y < end; y++ {
		line := s.line(y)
		for x := range line {
			line[x] = s.cursor.Attrs
		}
	}

	if how == 0 || how == 1 {
		s.EraseInLine(how, false)
	}
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Vt/Edit.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Vt;

public partial class Screen
{
    public void Draw(string data)
    {
        var cs = _charset == 1 ? _g1Charset : _g0Charset;
        foreach (var runeVal in data.EnumerateRunes())
        {
            var charCode = runeVal.Value;
            if (charCode >= 0 && charCode < 256)
            {
                charCode = cs[charCode];
            }

            var charWidth = Wcwidth.RuneWidth(charCode);

            if (_cursor.X == _columns)
            {
                if (_mode.ContainsKey(ModeCodes.Decawm))
                {
                    CarriageReturn();
                    LineFeed();
                }
                else if (charWidth > 0)
                {
                    _cursor.X -= charWidth;
                }
            }

            if (_mode.ContainsKey(ModeCodes.Irm) && charWidth > 0)
            {
                InsertCharacters(charWidth);
            }

            var line = Line(_cursor.Y);
            if (charWidth == 1)
            {
                line[_cursor.X] = VtHelpers.WithData(_cursor.Attrs, char.ConvertFromUtf32(charCode));
            }
            else if (charWidth == 2)
            {
                line[_cursor.X] = VtHelpers.WithData(_cursor.Attrs, char.ConvertFromUtf32(charCode));
                if (_cursor.X + 1 < _columns)
                {
                    line[_cursor.X + 1] = VtHelpers.WithData(_cursor.Attrs, "");
                }
            }
            else if (charWidth == 0 && Wcwidth.CombiningClass(charCode) > 0)
            {
                if (_cursor.X > 0)
                {
                    var last = VtHelpers.CellAt(line, _cursor.X - 1, DefaultChar());
                    var normalized = Normalize.NfcNormalize(last.Data + char.ConvertFromUtf32(charCode));
                    line[_cursor.X - 1] = VtHelpers.WithData(last, normalized);
                }
                else if (_cursor.Y > 0)
                {
                    var prev = Line(_cursor.Y - 1);
                    var last = VtHelpers.CellAt(prev, _columns - 1, DefaultChar());
                    var normalized = Normalize.NfcNormalize(last.Data + char.ConvertFromUtf32(charCode));
                    prev[_columns - 1] = VtHelpers.WithData(last, normalized);
                }
            }
            else
            {
                return;
            }

            if (charWidth > 0)
            {
                _cursor.X = Math.Min(_cursor.X + charWidth, _columns);
            }
        }
    }

    public void Index()
    {
        var (top, bottom) = MarginsOrScreen();
        if (_cursor.Y == bottom)
        {
            for (var y = top; y < bottom; y++)
            {
                if (_buffer.TryGetValue(y + 1, out var src))
                {
                    _buffer[y] = src;
                }
                else
                {
                    _buffer[y] = new Dictionary<int, Char>();
                }
            }

            _buffer.Remove(bottom);
        }
        else
        {
            CursorDown(0);
        }
    }

    public void ReverseIndex()
    {
        var (top, bottom) = MarginsOrScreen();
        if (_cursor.Y == top)
        {
            for (var y = bottom; y > top; y--)
            {
                if (_buffer.TryGetValue(y - 1, out var src))
                {
                    _buffer[y] = src;
                }
                else
                {
                    _buffer[y] = new Dictionary<int, Char>();
                }
            }

            _buffer.Remove(top);
        }
        else
        {
            CursorUp(0);
        }
    }

    public void LineFeed()
    {
        Index();
        if (_mode.ContainsKey(ModeCodes.Lnm))
        {
            CarriageReturn();
        }
    }

    public void InsertLines(int count)
    {
        if (count == 0)
        {
            count = 1;
        }

        var (top, bottom) = MarginsOrScreen();
        if (top <= _cursor.Y && _cursor.Y <= bottom)
        {
            for (var y = bottom; y >= _cursor.Y; y--)
            {
                if (y + count <= bottom)
                {
                    if (_buffer.TryGetValue(y, out var src))
                    {
                        _buffer[y + count] = src;
                    }
                }

                _buffer.Remove(y);
            }

            CarriageReturn();
        }
    }

    public void DeleteLines(int count)
    {
        if (count == 0)
        {
            count = 1;
        }

        var (top, bottom) = MarginsOrScreen();
        if (top <= _cursor.Y && _cursor.Y <= bottom)
        {
            for (var y = _cursor.Y; y <= bottom; y++)
            {
                if (y + count <= bottom)
                {
                    if (_buffer.TryGetValue(y + count, out var src))
                    {
                        _buffer.Remove(y + count);
                        _buffer[y] = src;
                    }
                }
                else
                {
                    _buffer.Remove(y);
                }
            }

            CarriageReturn();
        }
    }

    public void InsertCharacters(int count)
    {
        if (count == 0)
        {
            count = 1;
        }

        var line = Line(_cursor.Y);
        var def = DefaultChar();
        for (var x = _columns; x >= _cursor.X; x--)
        {
            if (x + count <= _columns)
            {
                line[x + count] = VtHelpers.CellAt(line, x, def);
            }

            line.Remove(x);
        }
    }

    public void DeleteCharacters(int count)
    {
        if (count == 0)
        {
            count = 1;
        }

        var line = Line(_cursor.Y);
        for (var x = _cursor.X; x < _columns; x++)
        {
            if (x + count <= _columns)
            {
                if (line.TryGetValue(x + count, out var src))
                {
                    line.Remove(x + count);
                    line[x] = src;
                }
                else
                {
                    line[x] = DefaultChar();
                }
            }
            else
            {
                line.Remove(x);
            }
        }
    }

    public void EraseCharacters(int count)
    {
        if (count == 0)
        {
            count = 1;
        }

        var line = Line(_cursor.Y);
        for (var x = _cursor.X; x < Math.Min(_cursor.X + count, _columns); x++)
        {
            line[x] = _cursor.Attrs;
        }
    }

    public void EraseInLine(int how, bool privateMode = false)
    {
        _ = privateMode;
        int start, end;
        switch (how)
        {
            case 0:
                start = _cursor.X;
                end = _columns;
                break;
            case 1:
                start = 0;
                end = _cursor.X + 1;
                break;
            case 2:
                start = 0;
                end = _columns;
                break;
            default:
                return;
        }

        var line = Line(_cursor.Y);
        for (var x = start; x < end; x++)
        {
            line[x] = _cursor.Attrs;
        }
    }

    public void EraseInDisplay(int how)
    {
        int start, end;
        switch (how)
        {
            case 0:
                start = _cursor.Y + 1;
                end = _lines;
                break;
            case 1:
                start = 0;
                end = _cursor.Y;
                break;
            case 2:
            case 3:
                start = 0;
                end = _lines;
                break;
            default:
                return;
        }

        for (var y = start; y < end; y++)
        {
            var line = Line(y);
            foreach (var x in line.Keys.ToList())
            {
                line[x] = _cursor.Attrs;
            }
        }

        if (how is 0 or 1)
        {
            EraseInLine(how, false);
        }
    }
}
